integration/image/delta_test.go

```diff
@@ -0,0 +1,373 @@
+		t.Fatal(err)
+	}
+
+	t.Log("Pushing delta to local registry")
+	encodedAuth, err := command.EncodeAuthToBase64(types.AuthConfig{
+		Username: reg.Username(),
+		Password: reg.Password(),
+		Auth:     "htpasswd",
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	rc, err = client.ImagePush(ctx, delta, types.ImagePushOptions{
+		RegistryAuth: encodedAuth,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	io.Copy(ioutil.Discard, rc)
+	rc.Close()
+
+	t.Log("Listing local images")
+	listImages(t, client)
+
+	t.Log("Stopping daemon")
+	d.Stop(t)
+
+	args = append(args, []string{
+		fmt.Sprintf("--delta-data-root=%s", d.Root),
+		fmt.Sprintf("--delta-storage-driver=%s", os.Getenv("DOCKER_GRAPHDRIVER")),
+	}...)
+	var newRootDir = fmt.Sprintf("%s/root-before", d.Folder)
+	d.Root = newRootDir
+
+	t.Log("Starting daemon with separate delta-data-root")
+	d.Start(t, args...)
+
+	t.Log("Pulling delta from local registry")
+	rc, err = client.ImagePull(ctx, delta, types.ImagePullOptions{
+		RegistryAuth: encodedAuth,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	parseCmdOutputForError(t, rc)
+	rc.Close()
+
+	hasImages(t, client, expectedImages)
+}
+
+func pullDeltaImages(t *testing.T, client apiclient.APIClient, base, target string) {
+	var (
+		err error
+		rc  io.ReadCloser
+		ctx = context.Background()
+	)
+
+	t.Log("Pulling delta base")
+	rc, err = client.ImagePull(ctx,
+		base,
+		types.ImagePullOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	io.Copy(ioutil.Discard, rc)
+	rc.Close()
+
+	t.Log("Pulling delta target")
+	rc, err = client.ImagePull(ctx,
+		target,
+		types.ImagePullOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	io.Copy(ioutil.Discard, rc)
+	rc.Close()
+}
+
+var digestRe = regexp.MustCompile("sha256:([a-fA-F0-9]{64})$")
+
+func parseDeltaImageID(t *testing.T, deltaCmdOutput io.Reader) (digest string) {
+	sc := bufio.NewScanner(deltaCmdOutput)
+	for sc.Scan() {
+		var v map[string]interface{}
+		if err := json.Unmarshal(sc.Bytes(), &v); err != nil {
+			t.Fatal(err)
+		}
+		if _, ok := v["errorDetail"]; ok {
+			t.Fatal(errors.New(v["errorDetail"].(string)))
+		}
+		_, ok := v["progressDetail"]
+		if ok {
+			continue
+		}
+		s, ok := v["status"]
+		if !ok {
+			continue
+		}
+		status, ok := s.(string)
+		if !ok {
+			continue
+		}
+		r := digestRe.FindStringSubmatch(status)
+		if len(r) < 1 {
+			continue
+		}
+		digest = r[0]
+	}
+	if err := sc.Err(); err != nil {
+		t.Fatal(err)
+	}
+	if digest == "" {
+		t.Fatal("Unable to parse delta image id from progress output")
+	}
+	return digest
+}
+
+func hasImages(t *testing.T, client apiclient.APIClient, contains []string) (ok bool) {
+	var (
+		err  error
+		list []types.ImageSummary
+		ctx  = context.Background()
+	)
+
+	list, err = client.ImageList(ctx, types.ImageListOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	var foundImages []string
+	for _, im := range list {
+		foundImages = append(foundImages, im.ID)
+	}
+
+	for _, im := range contains {
+		assert.Contains(t, foundImages, im)
```

Comment:
The function name `hasImages` and its `bool` return type sound like the function returns `true` or `false`, but as implemented it never returns `false` (if `assert` fails it raises an exception, right?). For increased clarity, I suggest replacing the `assert` with `return false`, then changing the function call from `hasImages(...)` to `assert.True(hasImages(...))`.

Reply:
connects to https://github.com/balena-os/balena-engine/pull/121#discussion_r233808140

I will change `hasImages` accordingly :)